src/patchelf.cc

```diff
@@ -1885,9 +1906,186 @@ void ElfFile<ElfFileParamNames>::clearSymbolVersions(const std::set<std::string>
+    {
+        return lower_bound(i)->second;
+    }
+
+    template <class T>
+    void incref(const char *idx, T *sym)
+    {
+    	(*this)[idx].push_front(sym);
+    }
+
+};
+
+template<ElfFileParams>
+void ElfFile<ElfFileParamNames()>::cleanstrtab()
+{
+    auto shdrDynamic = findSectionHeader(".dynamic");
+    auto shdrDynStr = findSectionHeader(".dynstr");
+    auto shdrDynSym = findSectionHeader(".dynsym");
+
+    auto strTab = (char *)fileContents->data() + rdi(shdrDynStr.sh_offset);
+    int verneednum = 0;
+    int verdefnum = 0;
+    //mkref(strTab, rdi(shdrDynStr.sh_size)); // new strInfo()
+    mymap m(strTab, rdi(shdrDynStr.sh_size));
+
+    Elf_Dyn* dyn = (Elf_Dyn *) (fileContents->data() + rdi(shdrDynamic.sh_offset));
+
+    /* NAMEs in .dynamic */
+    for (; rdi(dyn->d_tag) != DT_NULL; dyn++) {
+        switch (rdi(dyn->d_tag)) {
+        case DT_NEEDED: case DT_SONAME:
+        	m.incref(strTab + rdi(dyn->d_un.d_val), &dyn->d_un.d_val);
+
+            //incref(strTab + rdi(dyn->d_un.d_val), &dyn->d_un.d_val, sizeof dyn->d_un.d_val);
+            //m[strTab + rdi(dyn->d_un.d_val)].push_front(lsym);}
+        	break;
+        case DT_VERNEEDNUM:
+        	verneednum = rdi(dyn->d_un.d_val);
+        	break;
+        case DT_VERDEFNUM:
+        	verdefnum = rdi(dyn->d_un.d_val);
+        	break;
+        }
+    }
+    /* NAMEs in dynsym */
+    Elf_Sym* sym = (Elf_Sym *) (fileContents->data() + rdi(shdrDynSym.sh_offset));
+    int count = rdi(shdrDynSym.sh_size) / sizeof(Elf_Sym);
+    for (int i = 0; i < count; i++) {
+        auto dynsym = sym[i];
+        //incref(strTab + rdi(dynsym.st_name), &dynsym.st_name, sizeof dynsym.st_name);
+        m.incref(strTab + rdi(dynsym.st_name), &dynsym.st_name);
+    }
+
+    if (verneednum) {
+        auto shdrVerSymr = findSectionHeader(".gnu.version_r");
+ 		Elf_Shdr & shdrVersionRStrings = shdrs.at(rdi(shdrVerSymr.sh_link));
+		// this is where we find the actual version strings; is it dynstr?
+		if (shdrVersionRStrings.sh_offset == shdrDynStr.sh_offset) {
+			auto verr = fileContents->data() + rdi(shdrVerSymr.sh_offset);
+			auto versym = (Elf_Verneed *)verr;
+			while (verneednum>0) {
+				// NAMEs in verneed referring to dynstr
+				//incref(strTab + rdi(versym->vn_file), &versym->vn_file, sizeof versym->vn_file);
+				m.incref(strTab + rdi(versym->vn_file), &versym->vn_file);
+				int auxnum = rdi(versym->vn_cnt);
+				auto auxsym = (Elf_Vernaux *)((char*)versym + rdi(versym->vn_aux));
+				while (auxnum>0) {
+					//incref(strTab + rdi(auxsym->vna_name), &auxsym->vna_name, sizeof auxsym->vna_name);
+					m.incref(strTab + rdi(auxsym->vna_name), &auxsym->vna_name);
+					auxsym = (Elf_Vernaux *)((char*)auxsym + rdi(auxsym->vna_next));
+					auxnum--;
+				}
+				versym = (Elf_Verneed *)((char*)versym + rdi(versym->vn_next));
+				verneednum--;
+			}
+		}
+    }
+
+    if (verdefnum) {
+        auto shdrVerSymd = findSectionHeader(".gnu.version_d");
```

Comment:
Mixing tab and spaces. You might want to look into the identation to match the rest of the file

Reply:
I was hoping eclipse would do that automatically, but alas it doesn't... What is the style? tabs or spaces?

Reply:
Spaces